core/src/test/java/brooklyn/test/entity/RecordingSensorEventListener.java

```diff
@@ -0,0 +1,137 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package brooklyn.test.entity;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+import com.google.common.base.Objects;
+import com.google.common.base.Supplier;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+
+import brooklyn.entity.Entity;
+import brooklyn.event.Sensor;
+import brooklyn.event.SensorEvent;
+import brooklyn.event.SensorEventListener;
+
+public class RecordingSensorEventListener<T> implements SensorEventListener<T> {
+    
+    private final boolean skipDuplicateValues;
+    private final List<SensorEvent<T>> events = Collections.synchronizedList(Lists.<SensorEvent<T>>newArrayList());
+    private final List<T> values = Collections.synchronizedList(Lists.<T>newArrayList());
+    private final Multimap<EntityAndSensor<?>, SensorEvent<T>> eventsBySource = 
+            Multimaps.synchronizedListMultimap(ArrayListMultimap.<EntityAndSensor<?>, SensorEvent<T>>create());
+    private final Multimap<EntityAndSensor<?>, T> valuesBySource = 
+            Multimaps.synchronizedListMultimap(ArrayListMultimap.<EntityAndSensor<?>, T>create());
+
+    public RecordingSensorEventListener() {
+        this(false);
+    }
+    
+    public RecordingSensorEventListener(boolean skipDuplicateValues) {
+        this.skipDuplicateValues = skipDuplicateValues;
+    }
+    
+    @Override
+    public void onEvent(SensorEvent<T> event) {
+        EntityAndSensor<?> source = new EntityAndSensor<T>(event.getSource(),  event.getSensor());
+        Collection<T> previousValues = valuesBySource.get(source);
+        if (skipDuplicateValues && !previousValues.isEmpty() && Objects.equal(Iterables.get(previousValues, previousValues.size()-1), event.getValue())) {
+            // skip
+        } else {
+            events.add(event);
+            values.add(event.getValue());
+            eventsBySource.put(source, event);
+            valuesBySource.put(source, event.getValue());
+        }
+    }
+    
+    public void clear() {
+        events.clear();
+        eventsBySource.clear();
+        valuesBySource.clear();
+    }
+
+    public List<SensorEvent<T>> getEvents() {
+        return ImmutableList.copyOf(events);
+    }
+    
+    public List<T> getValues(Entity entity, Sensor<?> sensor) {
+        return ImmutableList.copyOf(valuesBySource.get(new EntityAndSensor(entity, sensor)));
+    }
+    
+    public List<T> getAllValues() {
+        return ImmutableList.copyOf(values);
+    }
+    
+    public Supplier<List<SensorEvent<T>>> getEventsSupplier() {
```

Comment:
Seemed an odd thing to include until I spotted `Asserts.eventually`, which takes a `Supplier`. Definitely worth a `@see` link.